HARJA-1691

solita-staskila opened the pull request:
- Otettu osioiden "cleanaus" pois kun tallennetaan, jotta sivu ei pomppisi 

#### Tehtävä- määrämuutos:
- Ennen 2025 alkaneet -> anna kirjata tavoitehinnan muutos aina käsin
- 2025 alkaneet -> täysi automaatio 
- Taulukko näkyy pelkästään 2025 alkaen 
- Sivutus poistettu kokonaan (design)

#### Testidata & testit
- Siirretty Kajaanin mhu:lle (2028 ->)

solita-staskila committed:
Poista sivutus, korjaa sivun pomppiminen tallentaessa

solita-staskila committed:
Anna kirjata tehtävämäärät käsin vanhemmille urakoille

solita-staskila committed:
Siirrä testidata MHU Kajaanille

solita-staskila committed:
Muuta yksikkötestejä kajaanin urakalle tehtävämäärissä

solita-staskila committed:
Korjaa termit, mhu+? -> uusi-urakka?

solita-staskila committed:
Korjaa analytiikka kuluid shift

solita-staskila committed:
Merge branch 'develop' into HARJA-1691

# Conflicts:
#	src/cljs/harja/tiedot/urakka/muutokset/yhteiset_tiedot.cljs

solita-staskila committed:
Korjaile merge konfliktit tuck stateen, lisää uusi-urakka parametri

K1ll3rF0x commented on src/clj/harja/kyselyt/muutos_kyselyt.sql:
```diff
@@ -518,13 +518,14 @@ SELECT
   CASE 
     -- ============================================================
     -- Yksikköhinta puuttuu -> muutos syötetään käsin 
+    -- Myös >2025 urakat syöttävät aina käsin
```
Tarkoitatko kuitenkin, että ennen 2025 alkaneet urakat syöttävät aina käsin <2025?

solita-staskila replied:
Joo just, väärinpäin siis väkänen, sitä tarkoitat? 

Ennen 2025 alkaneet urakat syöttävät aina tavoitehinnan muutoksetn käsin. se on nyt viimeisin tieto.

2025 alkavat -> täysi automatiikka

K1ll3rF0x commented on tietokanta/testidata/muutos_testidata.sql:
```diff
@@ -406,150 +479,187 @@ BEGIN
-    '[Muutokset] Määrämitattava toteuma 3');
+        (v_kayttaja_id,
+         (SELECT id FROM toteuma WHERE lisatieto = '[Muutokset] Määrämitattava toteuma 3 ' || p_urakka),
+         to_timestamp((p_vuosi - 2)::text || '-11-30 17:00:00', 'YYYY-MM-DD HH24:MI:SS'),
```
Kun käytetään tällaisia offsettejä, niin testidatan generoivan funktion käyttäjän pitää hoksata antaa riittävän iso p_vuosi, jotta toteumat/kulut/yms. ei mene Urakan vuosirajojen ulkopuolelle.
Tästä voisi hoksauttaa jossakin kommenteissa ainakin, ellei asiaan ota kantaa ohjelmallisesti esim. heittämällä virhettä kun:
p_vuosi - 2 < urakan_alkuvuosi

solita-staskila replied:
Joo, laadin tuohon jotain, ihan hyvä pointti.

Tuo ei ole kovin helposti ymmärrettävissäkään, mutta se on kuitenkin vaan testidataa.

Toi asettaa halutulle vuodelle datan sellaiseksi että yksikköhinnan valinnat näkyy tietyillä tehtävillä, ja sitä voi suoraa testiharjassa naputella.

solita-staskila replied:
Laitoin raise exceptionit, migraation ajo stoppaa jos syötetty vuosi ei osu haarukkaan

K1ll3rF0x commented on src/clj/harja/kyselyt/muutos_kyselyt.sql:
```diff
@@ -518,13 +518,14 @@ SELECT
+    -- Ei jatketa myöskään, jos kyseessä vanhempi urakka
     -- ============================================================
-    WHEN SUM(urakan_tehtavat.maara) = 0 THEN NULL
+    WHEN (:uusi-urakka? = false OR (SUM(urakan_tehtavat.maara) = 0)) THEN NULL
```
:uusi-urakka? nimi voisi olla selkeämpi:
Käsittääkseni tätä optiota käytetään kauttaaltaan muutoksissa vain siihen, että onko manuaalisyöttö päällä vai ei.
Option nimi voisi viitata siis siihen ihteensä, eli onko täysi automatiikka päällä/pois.

solita-staskila replied:
Vaihdoin ton -> laskenta-automatiikka?

K1ll3rF0x commented on src/cljs/harja/views/urakka/muutokset/lasketut_muutokset.cljs:
```diff
@@ -102,15 +102,9 @@
          [grid/grid
           {:tunniste :id
-           ;; Annetaan tälle sivutus, voi olla paljon tehtäviä 
-           :sivuta 20
```
Miten sivutuksen poisto vaikuttaa suorituskykyyn?
Onko testattu isolla tehtävämäärällä, joka voisi vastata todellista tuotannon tilannetta?

solita-staskila replied:
Voi olla jotain tosi pientä front vaikutusta, mutta yhtä paljon elementtejä  siihen renderöidään ja haetaan sivutuksellakin.

Tähän oli Designerit alusta asti sitä mieltä, että ei sivutusta.
Minä sen siihen alunperin laitoin, jotta olisi helpompi käyttää isoa taulukkoa.

Tässä kuitenkin ongelmia: 

- Yksikköhinta valitaan -> sivutus nollaantuu
- Taulukon yhteenvetorivi näyttää euroja, jota ei ole valitulla sivulla -> aiheuttaa hämmennystä
- Jos tätä käytetään rahojen siirtoon tehtävältä toiselle, ei näe siirrettäviä tehtäviä samalla sivulla 
- Väliotsikko voi jäädä myös edelliselle sivulle, jos sivutus on käytössä 

Joten, otetaan sivutus vaan pois. 

Pyöräytän tämän kloonikannalle, jos saan infran konffattua, ja jos saan dataa sinne syötettyä. 
Tai, jos jopa löytyy valmis urakka. Tsekkaan tämän miten siellä toimii.

solita-staskila committed:
Korjaa desimaalilukuja, näytä vain tarvittaessa

solita-staskila committed:
Korjaile lomakkeen lataus, disablointi, virheviestit

solita-staskila committed:
Osion näkyminen ja cleanaus tietojen haussa

Parannettu jälleen kuinka näkymä käyttäytyy kun tietoja tallentaa, tai näkymään tullaan. Vähemmän pomppimista, oikeat loaderit, ja näin ja noin.

solita-staskila committed:
Selvennä termi laskennan automatiikasta

solita-staskila committed:
Lisää testidataan raise jos vuosi invalid

solita-staskila committed:
Älä näytä muutokset näkymää ennen 2025

solita-staskila committed:
Näytä määrämuutoksessa plusmerkki

solita-staskila committed:
Mahdollista yksikköhinnan valinnan poisto

Lisää 'Ei yksikköhintaa' valinta modaliin, sekä aseta kantaan oikeat tiedot (versiointi tarkastettu myös)

solita-staskila committed:
Korjaa näkymän pomppiminen yksikköhinnassa, poista turha haku

solita-staskila committed:
Korjaa miinus arvojen muokkaus gridissä.. huoh

solita-staskila committed:
Laske yksikköhinta manuaalisesti asetetulla muutoksella

K1ll3rF0x commented on src/cljs/harja/ui/kentat.cljs:
```diff
@@ -379,9 +380,11 @@
               nykyinen-teksti (or @teksti
-                                  (normalisoi-numero (fmt nykyinen-data) salli-whitespace?)
-                                  "")
-              kokonaisluku-re-pattern (re-pattern (str "-?\\d{1," kokonaisosan-maara "}"))
```
Kannattaa huomioida, että täällä on tulossa kokonaisvaltainen korjaus :numero-kenttään ja myös näiden miinusarvojen käsittelyyn:

https://github.com/finnishtransportagency/harja/pull/3994/files#diff-531afb38392a6ddea98ad45b4f89ae12a9a0d1845eae1e453863601c0d2f9140

Ei ehditty tätä Reagent-muutosta mergettää, kun testaaminen jäi kesken ja sitten tuli vielä työryhmän toisella jäsenellä lomat päälle.
Nämä Reagent-branchissä olevat korjaukset tulee olemaan se lopullinen ratkaisu, mitä tullaan jatkossa käyttämään. 

Sanon tämän siksi, ettet käytä liikaa aikaa tuon :numero-kentän ongelmien korjailuun.

K1ll3rF0x replied:
Reagent-muutokset toki saataisiin mergettyä, jos joku ehtisi niiden testaamisessa ja korjaamisessa auttaa.

Siellä oli yksi ongelma vielä korjaamatta:
https://github.com/finnishtransportagency/harja/pull/3994#issuecomment-3438215628

Muuten on loppusuoralla. Mutta, ehkä palaamme tähän ensi viikolla.

solita-staskila replied:
Otin tuolta muutokset tähän haaraan. Näytti toimivan ihan tarpeeksi hyvin.

note:
Jos arvo on -25,00 
-> Lähde kumittamaan 5 kohdalta
-> Ei anna kumittaa numeroa 2

solita-staskila committed:
Merge branch 'develop' into HARJA-1691

K1ll3rF0x commented on src/cljs/harja/ui/ikonit.cljs:
```diff
@@ -971,11 +971,16 @@
+         :alt (name ikoni)
+         :width (str (or koko 24) "px")
+         :style (cond
+                  (= ikoni :alert)
```
Voisiko tämän skaalausongelman korjata suoraan ikoniin, niin ei tarvitse alkaa skaalailemaan koodissa. Se on hyvin simppeli korjata suoraan svg:hen asettamalla sinne 0.8 skaalauskerroin, niin sinne ikonin ympärille tulee hyvin ilmaa ympärille.

solita-staskila replied:
Joo, korjattu 

https://github.com/finnishtransportagency/harja/pull/4129/commits/4fbeeb61af4b53caa58383faa2db4d99c89aa179

K1ll3rF0x reviewed (commented):
Reviewin myötä joitakin kommentteja lisätty ja parannusehdotuksia, joista suurin osa on jo tehtykin.
Vielä testailua ja sitten olisi PR merge-kunnossa. Koodista en löytänyt isompaa kommentoitavaa enää.

Reagentin päivitysbranchissä on muistettava merge-konflikteja korjatessa korvata :numero-kenttään tehdyt muutokset Reagent-branchin ratkaisulla. kun develop branch mergetään kyseiseen branchiin tämän PR:n mergen jälkeen.

solita-staskila committed:
Merge branch 'develop' into HARJA-1691

solita-staskila committed:
Älä tee turhia kutsuja, jos sivua ei näytetä

Jos valittu hoitokausi on alle 25, älä hae muutostöitä ollenkaan

solita-staskila committed:
Enabloi kulujen muutoshaku kehitysympäristöön

solita-staskila committed:
Ota reagent haarasta kentat kopio miinusmerkkien korjaukseen

solita-staskila committed:
Skaalaa svg ikoni svg:ssä suoraa

solita-staskila committed:
Kopioi reagent haaran kentat ilman formatointia

solita-staskila committed:
Poista vanhentunut testi

Testin ajo ei onnistunut joten en käytä aikaa vanhentuneen testiframeworkin korjailuun

solita-staskila committed:
Näytä rahavaraus taulukossa nollat default

solita-staskila committed:
Formatoi pysyvän muutokset desimaalit

solita-staskila committed:
Korjaa cypy testi uuteen normalisoituun miinusmerkkiin

solita-staskila committed:
Korjaile määrämitattavien tehtävien näkyvyystä

Aluetieto/mhu tehtava ei tarvitse, tuo maaramitattava? korvaa nämä

solita-staskila committed:
Korjaa rahavarausten takaend testejä

solita-staskila committed:
Formatoi määrämutos-sarake, korjaa testejä

solita-staskila committed:
Näytä tav hinnan muutoksessa myös plus arvot

solita-staskila committed:
Revert "Korjaa rahavarausten takaend testejä"

This reverts commit 733406b211239031010a4be2acead58c2c0b9fa0.

solita-staskila committed:
Merge branch 'develop' into HARJA-1691

K1ll3rF0x commented on test/cljs/harja/ui/kentat_test.cljs:
```diff
@@ -68,12 +68,6 @@
      (is (= 80 @data))
 
-     "Keskeneräinen numero ei päivitä dataa"
-     (val! "-")
```
Tän testin vois jättää tähän, vaikka pois kommentoituna jos haittaa muutoin.
Se on tärkeä testi tuossa uudessa Reagent-versiossa ja sen myötä tehdyissä muutoksissa.

Kyseisestä testistä variantti tulossa täällä:
[‎test/cljs/harja/ui/kentat_test.cljs‎](https://github.com/finnishtransportagency/harja/pull/3994/files#diff-715c362a768bcf2b8c8ba78bf8d75ad632fd7d52b8731f1e1afa27ac27938082)

K1ll3rF0x reviewed (commented):
Yksi muutospyyntö liittyen kentat_test.cljs.
Sen jälkeen merge 🚀

K1ll3rF0x committed:
Merge branch 'develop' into HARJA-1691

solita-staskila committed:
Lisää pHant0m testi kenttiin
